[#3, #4] 계정 정보 데이터 작업 기능 구현

donsonioc2010 opened the pull request:
계정 데이터에 대한 조회 및DB 처리 기능입니다.

작업을 진행하다 보니 같은 브랜치에서 같이 진행을 해버렸습니다...


기능의 경우에는 Class를 분리해서 작성하였습니다.
[#3] 기능들의 경우에는 SummonerAccount로 명칭을 칭하였고
[#4] 기능들의 경우에는 SummonerHistory로 명칭을 칭하였습니다.

donsonioc2010 committed:
feat : 매일마다 갱신되는 API키 업데이트를 위한 기능 변경

donsonioc2010 committed:
feat : 소환사 명을 검색할경우 검색결과 Return기능 구현

donsonioc2010 committed:
feat : 사용자의 ID값을 토대로 랭크게임 정보를 획득하는 기능 구현

donsonioc2010 committed:
added : yml설정파일 부재로 인한 테스트 오류로 추가

donsonioc2010 committed:
edited : memory방식 설정 수정

donsonioc2010 commented on src/test/java/com/nooblol/account/service/impl/SummonerHistoryServiceImplTest.java:
```diff
@@ -0,0 +1,47 @@
+import org.springframework.test.context.ActiveProfiles;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
```
해당 테스트클래스의 데이터가이 결국 Riot에서 가져온 데이터를 기반으로 테스트케이스를 진행을 해야하는데
데이터들이 불변성이 있는 데이터가 없는 것들이 더 많아다보니 다음과 같은 생각을 많이 하였습니다.

1.  지속적으로 변화하는 것에 대하여 테스트케이스를 제작을 해야하나?
2. 데이터가 지속적으로 변하는 상황에서 테스트 케이스를 어떻게 작성해야 하지?
3. 일단 생각해 볼 수 있는건 고정적인 값인 소환사계정의 ID값밖에 없다.

donsonioc2010 commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+    }
+    return summonerHistoryProcess(summonerId, sync);
+  }
+
```
[#4] 의 기능을 제작한 부분입니다.
sync를 통해서 DB의 값을 가져갈지 또는 Riot서버의 값을 가져갈지를 기능으로 구현했는데
좀더 보기 좋게 작성할 수 있는 방버이 더 보이질 않았습니다..

f-lab-kyle commented on .github/workflows/ci_test.yml:
```diff
@@ -31,6 +31,15 @@ jobs:
           distribution: 'temurin'
 
+      ## Riot Api Key 수정
+      - name: Constants.yml Change Riot Api Key
+        run: |
+          cd ./src/main/resources
+          chmod 755 constants.yml
+          sed -i 's/riot-api-key/${{ secrets.RIOT_API_KEY }}/g' constants.yml 
+          cat constants.yml
+        shell: bash
```
api key 를 application 빌드(혹은 실행) 시점에 주입시켜주는 것으로 이해 했는데요.
이렇게 해도 크게 상관은 없으나, application이 `github > secrets` 이라는 연결고리가 하나 더 생기는것 같은 느낌이 드네요 ㅎ
그리고 이전 미팅때 말씀 드린것처럼 서버를 실행한 다음에는 반영이 안되는 문제도 있을것 같아요 (즉, api key가 만료되면 갱신/재시작을 하지 않는 이상엔 application이 정상동작하지 않을것 같거든요)

제가 이해한 요구사항은 아래와 같은데요.
1. api key를 동적으로 바꿔주고 싶다.
2. 서버를 재시작 하지 않아도 동적으로 바꿔주고 싶다.

api key의 만료 시점은 알고 있으니 갱신을 자동화 하는방법을 찾아보고 (새로운  api key 발급을 홈페이지에서 클릭하는 것이 아닌 프로그램으로 자동화 획득)
이를 db에 넣고 ( 혹은 지난번에 이야기 했던 remote config server 활용 ) 이를 갱신이 된 다음에 다시 셋팅 (메모리에?) 하면 될것 같네요.

f-lab-kyle commented on src/main/java/com/nooblol/account/controller/SummonerController.java:
```diff
@@ -0,0 +1,37 @@
+
+@RestController
+@Controller
+@RequestMapping("/summoner/*")
```
`@RequestMapping("/summoner")` 이렇게 하지 않은 이유가 무엇일까요?

f-lab-kyle commented on src/main/java/com/nooblol/account/controller/SummonerController.java:
```diff
@@ -0,0 +1,37 @@
+    return summonerService.getSummonerAccointInfo(summonerName);
+  }
+
+  @GetMapping("/history")
```
@GetMapping("/history")
@GetMapping("history")

이 둘의 차이가 무엇일까요? 일관되게 작성하는게 좋을 것 같아요.

f-lab-kyle commented on src/main/java/com/nooblol/account/controller/SummonerController.java:
```diff
@@ -0,0 +1,37 @@
+public class SummonerController {
+
+  @Autowired
+  SummonerService summonerService;
```
클래스 변수의 접근제한자가 없어보입니다.

f-lab-kyle commented on src/main/java/com/nooblol/account/controller/SummonerController.java:
```diff
@@ -0,0 +1,37 @@
+@RequestMapping("/summoner/*")
+public class SummonerController {
+
+  @Autowired
```
bean 을 주입하는 방법중에 Autowired 을 사용한 이유가 무엇일까요?

f-lab-kyle commented on src/main/java/com/nooblol/account/mapper/SummonerHistoryMapper.java:
```diff
@@ -0,0 +1,17 @@
+@Mapper
+public interface SummonerHistoryMapper {
+
+  List<SummonerHistoryDto> selSummonerHistoryById(String summonerId);
```
혹시 sel 이 select의 약자 일까요?
그게 맞다면 가급적 약자는 사용하지 않는게 좋을듯해요. 약자를 사용해서의 이점은 코딩을 할때 타이핑을 줄이는것 이외에 없을것 같아요. 오히려 가독성이 떨어져 보입니다.

아래 upd, ins 이런 녀석들도요.

f-lab-kyle commented on src/main/java/com/nooblol/account/mapper/SummonerHistoryMapper.java:
```diff
@@ -0,0 +1,17 @@
+
+  List<SummonerHistoryDto> selSummonerHistoryById(String summonerId);
+
+  SummonerHistoryDto selSummonerHistoryByLeagueAndId(SummonerHistoryDto summonerHistoryDto);
```
SummonerHistoryDto 이 변수를 전부 사용하는게 아니라면 정말 사용하는 변수만 넘기는게 좋을것 같아요.

f-lab-kyle commented on src/main/java/com/nooblol/account/controller/SummonerController.java:
```diff
@@ -0,0 +1,37 @@
+  SummonerHistoryServiceImpl summonerHistoryService;
+
+  @GetMapping("search/name")
+  public ResponseDto searchSummonerByName(@RequestParam(value = "summonerName") String summonerName)
```
주석이 하나도 안보이는데요.
쓸데 없는 내용의 주석은 지우는게 맞지만, 메소드/클래스/변수 들의 역할들에 대해서는 주석을 남겨두는게 혼동이 없어 보입니다.
이는 협업시에도 중요한 부분일것 같아요.

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+
+  @Override
+  public ResponseDto getSummonerHistoryInfo(String summonerId, boolean sync) {
+    if (summonerId.isBlank()) {
```
https://www.baeldung.com/javax-validation 요런것도 활용해보세요.

f-lab-kyle replied:
만약에 summonerId가 null이면 어떻게 될까요~?

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+    return summonerHistoryProcess(summonerId, sync);
+  }
+
+  ResponseDto summonerHistoryProcess(String summonerId, boolean sync) {
```
접근제한자가 없네요.

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+      List<SummonerHistoryDto> dbSummonerHistoryList =
+          summonerHistoryMapper.selSummonerHistoryById(summonerId);
+      if (dbSummonerHistoryList.isEmpty()) {
+        sync = false;
```
false 로 변경해주는 이유가 뭘까요?

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+        sync = false;
+      } else {
+        return new ResponseDto(HttpStatus.OK.value(), dbSummonerHistoryList);
+      }
```
```suggestion
      }
      return new ResponseDto(HttpStatus.OK.value(), dbSummonerHistoryList);
```

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+    } catch (IOException e) {
+      e.printStackTrace();
+    } catch (Exception e) {
+      e.printStackTrace();
```
1. 이렇게 되면 로그는 어디에 남나요?
2. 로그를 남길땐 최대한 로그가 발생했을때의 상황(정보)를 많이 남겨둬야 추후 로그를 볼때 유익(?) 합니다.

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+    } catch (Exception e) {
+      e.printStackTrace();
+    } finally {
+      if (CommonUtils.objectIsNull(rtnDto)) {
+        rtnDto = new ResponseDto(HttpStatus.NOT_FOUND.value(), HttpStatus.NOT_FOUND);
+      }
```
```suggestion
      if (CommonUtils.objectIsNull(rtnDto)) {
        return new ResponseDto(HttpStatus.NOT_FOUND.value(), HttpStatus.NOT_FOUND);
      }
```

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  private HttpResponse getApiResponseData(String url) throws IOException {
+    HttpClient client = HttpClientBuilder.create().build();
```
api를 사용할때마다 매번 HttpClient를 생성해야만 할까요~?

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+  private ArrayList<SummonerHistoryDto> getResponseBodyToDto(HttpResponse response)
+      throws IOException {
+    ResponseHandler<String> handler = new BasicResponseHandler();
+    ObjectMapper objectMapper = new ObjectMapper();
```
자주 사용하는 객체들은 매번 생성하지 말고 bean으로 관리하는게 효율적일것 같아요.

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+    list = objectMapper.readValue(body, new TypeReference<ArrayList<SummonerHistoryDto>>() {
+    });
+
+    return list;
```
```suggestion
    String body = handler.handleResponse(response);

    return objectMapper.readValue(body, new TypeReference<ArrayList<SummonerHistoryDto>>() {
    });
```

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+    ArrayList<SummonerHistoryDto> summonerHistoryList =
+        (ArrayList<SummonerHistoryDto>) responseDto.getResult();
+
+    summonerHistoryList.forEach(
```
summonerHistoryList 가 null일 경우는 없을까요?

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+    if (summonerHistoryDto.getLeagueId().isEmpty() ||
+        summonerHistoryDto.getSummonerId().isEmpty()) {
+      throw new NullPointerException("LeagueId or Summoner ID Is Null");
+    }
```
NullPointerException을 유발하는것보다 입력값이 잘못되었다는 예외가 더 어울려 보이구요.
string 이 비어있는 체크는 StringUtils 같은걸 활용해보세요. npe 처리도 해준답니다.

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+      summonerHistoryMapper.updSummonerHistry(summonerHistoryDto);
+    } else {
+      summonerHistoryMapper.insSummonerHistory(summonerHistoryDto);
+    }
```
upsert ( ON DUPLICATE KEY UPDATE ... ) 같은걸 구현하면 어떨까요?

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+
+@Service
+@RequiredArgsConstructor
+public class SummonerServiceImpl implements SummonerService {
```
왜 SummonerService을 바로 사용하지 않고 SummonerService 을 interface로 만들고 implements 받는 구조로 만들었나요?

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+  @Override
+  public ResponseDto selSummonerAccountByRiot(String summonerName) {
+    summonerName = summonerNameWhiteSpaceReplace(summonerName);
+    String url = riotConfiguration.getSummonerNameSearchApiUrl() + "by-name/" + summonerName;
```
path 도 url을 관리하는 식으로 동일하게 관리하면 어떨까요~?

f-lab-kyle commented on src/main/java/com/nooblol/global/utils/CommonUtils.java:
```diff
@@ -0,0 +1,20 @@
+    return name.replaceAll(" ", "%20");
+  }
+
+  public static Boolean objectIsNull(Object obj) {
+    return obj == null;
+  }
+
+  public static Boolean objectIsNotNull(Object obj) {
+    return obj != null;
+  }
```
메소드로 추출한 이유가 무엇일까요?

f-lab-kyle commented on src/main/resources/application-local.yml:
```diff
@@ -1,7 +1,8 @@
 spring:
   datasource:
-    url: jdbc:h2:tcp://localhost/~/h2/nooblol
+    #    url: jdbc:h2:mem://localhost/~/h2/nooblol
```
주석을 남겨놓는 이유가 있을까요~?

f-lab-kyle commented on src/main/resources/constants.yml:
```diff
@@ -1,5 +1,5 @@
-
+  apiKey: riot-api-key
+  summoner-name-search-api-url: https://kr.api.riotgames.com/lol/summoner/v4/summoners/
+  summoner-history-search-api-url: https://kr.api.riotgames.com/lol/league/v4/entries/by-summoner/
```
domain 과 path로 관리하는 편이 더 직관적일것 같습니다.

f-lab-kyle commented on src/main/resources/mybatis/mapper/account/SummonerMapper.xml:
```diff
@@ -0,0 +1,33 @@
+    SELECT
+    id, account_id, puuid, name, profile_icon_id, revision_Date, summoner_level
+    FROM SUMMONER_ACCOUNT
+    WHERE 1=1
```
가장 위험한 1=1
동적쿼리를 위한다면 차라리 <where> 를 알아보세요.
https://mybatis.org/mybatis-3/dynamic-sql.html#trim-where-set

f-lab-kyle commented on src/test/java/com/nooblol/account/service/SummonerServiceTests.java:
```diff
@@ -0,0 +1,48 @@
+  SummonerMapper summonerMapper;
+
+  @Test
+  @DisplayName("존재하는 소환사명 테스트")
```
테스트는 하나의 문장 형태가 보기 좋습니다. 테스트를 하고자 하는 조건과 목적이요.
예로들어, 존재하는 소환사를 조회하면 응답이 ok로 내려온다.

f-lab-kyle commented on src/test/java/com/nooblol/account/service/SummonerServiceTests.java:
```diff
@@ -0,0 +1,48 @@
+  @Test
+  @DisplayName("소환사명이 공백인 경우 Exception테스트")
+  void getSummonerAccountInfo_SummonerNameIsNull() {
+    try {
+      summonerService.getSummonerAccointInfo("");
+    } catch (IllegalArgumentException e) {
+      Assertions.assertEquals("소환사명이 입력되지 않았습니다", e.getMessage());
+    }
```
assertThrows

f-lab-kyle reviewed (commented):
고생하셨습니다! 

테스트 코드는 전반적으로 다시 보면 좋을것 같습니다. 
지금의 테스트 코드는 bean을 전부 띄우기 때문에 느릴것 같아요. 더불어 데이터를 직접 조회하니 테스트의 의존성이 데이터가 추가 되기도 하구요.
mock test에 대해서 알아보시는것도 좋을것 같습니다.

donsonioc2010 committed:
fix : 쿼리 1=1 제거 및 동적쿼리 수정

donsonioc2010 committed:
edited : 컨트롤러 Mapping경로 수정, 도메인과 path 분할

donsonioc2010 committed:
feat : 기능 확인을 할 수 있도록 log의 추가

donsonioc2010 committed:
chore : 로그 추가로 인한 gradle 수정

donsonioc2010 committed:
refactor : query명령 메소드 fullName으로 수정

donsonioc2010 committed:
test : 명칭 변경으로 인한 test수정

donsonioc2010 committed:
fix : log 경로 설정 오류 수정

donsonioc2010 committed:
refactor :
1. 문자열 StringUtils로 변경 및, 그에 따른 dependency추가,
2. exception 변경
3. sync관련 false변경문 삭제
4. summonerHistoryDto Null인경우 반복문 타지 않도록 수정

donsonioc2010 committed:
refactor : dto 사용하지 않는 항목들 제거

donsonioc2010 committed:
feat : ObjectMapper Custom 및 Bean등록

donsonioc2010 committed:
feat : HttpClient Bean을 통한 관리로 수정 및 그에 따른 소스코드 수정

donsonioc2010 committed:
refactor : Select Parameter Dto에서 사용하는 Parameter만 주는 것으로 수정

donsonioc2010 committed:
docs : 기능에 대하여 큰틀의 주석 추가

donsonioc2010 committed:
refactor : develop풀리퀘시에만 CI진행하도록 수정

donsonioc2010 committed:
refactor : CI진행시 PR출력되는 내용 테스트 결과만 나오도록 수정

donsonioc2010 committed:
test : Test Refactor And Mock Chang
1. Exception의 경우 assertThrown 변경
2. ServiceTest Mock으로 변경
3. DisplayName 문장형태로 변경
